### PR TITLE
Enhanced twg settings validation

### PR DESCRIPTION
**Pull request by kuasha420**
## Summary

Addresses issue:

- #5461 

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Modules/Thank_With_Google/Settings.php [hunk 1]
```diff
@@ -69,4 +69,103 @@ protected function get_default() {
...
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $button_post_types The value to sanitize.
+	 * @return array The sanitized value.
+	 */
+	private function sanitize_button_post_types( $button_post_types ) {
+		if ( ! is_array( $button_post_types ) ) {
+			return array( 'post' );
+		}
```
Comment (tofumatt) [inline includes/Modules/Thank_With_Google/Settings.php:168]: Why is this here? I guess it's for the default value/settings? I'd think a safer default is to show this on no post types by default 🤔 

```suggestion
		if ( ! is_array( $button_post_types ) ) {
			return array();
		}
```

File: includes/Modules/Thank_With_Google/Settings.php [hunk 2]
```diff
@@ -69,4 +69,103 @@ protected function get_default() {
...
+		return array_intersect( $button_post_types, get_post_types( array( 'public' => true ) ) );
```
Comment (tofumatt) [inline includes/Modules/Thank_With_Google/Settings.php:169]: Additionally, I wonder if this matches the UI we show… in the UI we only show posts that support `viewable` (see: https://github.com/google/site-kit-wp/issues/5456). I wonder if this would allow a user to submit a post type successfully that wouldn't appear in the UI.

Not a big deal as they'd need to intentionally do so, but it would be nice if the only posts we allowed to save would be ones that would appear with the `'viewable'` value in `supports`. 🤔 

I know this is what was in the ACs/IB, but I think it's maybe a bit inaccurate? Would there be an easy way to get all post types that match what `wp.data.select('core').getPostTypes().filter((postType) => postType.viewable)` returns? If so let's do that 😄

**Review by tofumatt — CHANGES_REQUESTED** [review]
Looks good, just one question and also a thought on the post types we allow to be saved.

Let me know what you think, after that I think we're good to merge 😄

File: includes/Modules/Thank_With_Google/Settings.php (hunk 2 above)
Comment (kuasha420) [reply includes/Modules/Thank_With_Google/Settings.php:169]: Hi, the API for `getPostTypes()` uses `is_post_type_viewable` for each post to determine the viewable status. For a post to be viewable: 

* if it's a builtin post type, the `public` must be `true`.
* If it's not a builtin post type, the `publicly_queryable` must be `true`. 

Which definitely makes your point valid. I'll see what's the best way to achieve that here. 

Off Topic: speaking of `getPostTypes` be aware that it only returns 10 post types if the site has more. See [here](https://github.com/WordPress/gutenberg/issues/15413). The workaround is to supply `{ per_page: -1 }` to the selector which will return 100. If there's 101 post type this workaround will not work :grin:

File: tests/phpunit/integration/Modules/Thank_With_Google/SettingsTest.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Class Google\Site_Kit\Tests\Modules\Thank_With_Google\Settings
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Modules/Thank_With_Google/SettingsTest.php:3]:
```suggestion
 * Class Google\Site_Kit\Tests\Modules\Thank_With_Google\SettingsTest
```

File: includes/Modules/Thank_With_Google/Settings.php [hunk 4]
```diff
@@ -69,4 +69,103 @@ protected function get_default() {
 			'buttonPostTypes' => array( 'post' ),
 		);
 	}
+
+	/**
+	 * Gets the setting value.
+	 *
+	 * @since m.e.x.t
```
Comment (aaemnnosttv) [inline includes/Modules/Thank_With_Google/Settings.php:76]: Please be sure to use the correct placeholder as search/replace will miss it otherwise.
```suggestion
	 * @since n.e.x.t
```

File: includes/Modules/Thank_With_Google/Settings.php (hunk 2 above)
Comment (tofumatt) [reply includes/Modules/Thank_With_Google/Settings.php:169]: Yeah, sadly I think that's our best option for now until someone runs into an issue 😅 

We could add pagination if needed, but for now `101` is probably enough.

File: includes/Modules/Thank_With_Google/Settings.php (hunk 1 above)
Comment (kuasha420) [reply includes/Modules/Thank_With_Google/Settings.php:168]: SGTM. :+1:

File: includes/Modules/Thank_With_Google/Settings.php (hunk 4 above)
Comment (kuasha420) [reply includes/Modules/Thank_With_Google/Settings.php:76]: Oh my goodness, they looks same to me on monospaced font. Sorry!

Comment (tofumatt) [reply includes/Modules/Thank_With_Google/Settings.php]: I missed it too, waiting for someone to make a `m.e.x.t` release in ZenHub just to really mess with us 😉

**Review by tofumatt — APPROVED** [review]
Code looks great, thanks! I'm just hunting down the _huge_ filesize increase for Thanks with Google before merging.

**Comment by tofumatt** [comment]
Update: looks like a bug, because I ran it locally and got this:
CleanShot 2022-07-21 at 19 26 46

:shipit:
